core/drivers/crypto/caam/hal/common/hal_ctrl.c

```diff
@@ -128,3 +130,127 @@ void caam_hal_ctrl_inc_priblob(vaddr_t baseaddr)
 	if (val != blob)
 		panic("Written PRIBLOB and read PRIBLOB do not match!");
 }
+
+uint8_t caam_hal_ctrl_get_mpcurve(vaddr_t ctrl_addr)
+{
+	uint32_t val_scfgr = 0;
+
+	/*
+	 * On i.MX8MQ B0, the MP is not usable, hence
+	 * return UINT8_MAX
+	 */
+	if (soc_is_imx8mq_b0_layer())
+		return UINT8_MAX;
+
+	/*
+	 * Verify if the device is closed or not
+	 * If device is closed, check get the MPCurve
+	 */
+	if (snvs_is_device_closed()) {
+		/* Get the SCFGR content */
+		val_scfgr = io_caam_read32(ctrl_addr + SCFGR);
+
+		/* Get the MPCurve field value - 4 bits */
+		val_scfgr = (val_scfgr & BM_SCFGR_MPCURVE) >> BS_SCFGR_MPCURVE;
+
+		/*
+		 * If the device is closed and the MPCurve field is 0
+		 * return UINT8_MAX indicating that there is a problem and the
+		 * MP can not be supported.
+		 */
+		if (!val_scfgr)
+			return UINT8_MAX;
+	}
+
+	return val_scfgr;
+}
+
+TEE_Result caam_hal_ctrl_read_mpmr(vaddr_t ctrl_addr, struct caambuf *mpmr)
+{
+	int i = 0;
+	uint32_t val = 0;
+
+	if (mpmr->length < MPMR_NB_REG) {
+		mpmr->length = MPMR_NB_REG;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+
+	/* MPMR endianness is reverted between write and read */
+	for (; i < MPMR_NB_REG; i += 4) {
+		val = io_caam_read32(ctrl_addr + MPMR + i);
+		mpmr->data[i] = (uint8_t)((val >> 24) & UINT8_MAX);
+		mpmr->data[i + 1] = (uint8_t)((val >> 16) & UINT8_MAX);
+		mpmr->data[i + 2] = (uint8_t)((val >> 8) & UINT8_MAX);
+		mpmr->data[i + 3] = (uint8_t)(val & UINT8_MAX);
+	}
+
+	mpmr->length = MPMR_NB_REG;
+	return TEE_SUCCESS;
+}
+
+bool caam_hal_ctrl_is_mp_set(vaddr_t ctrl_addr)
+{
+	return io_caam_read32(ctrl_addr + SCFGR) & BM_SCFGR_MPMRL;
+}
+
+void caam_hal_ctrl_fill_mpmr(vaddr_t ctrl_addr, struct caambuf *msg_mpmr)
+{
+	size_t i = 0;
+	vaddr_t reg = ctrl_addr + MPMR;
+	bool is_filled = false;
+	uint32_t val = 0;
+	size_t min_size = 0;
+	size_t remain_size = 0;
+
+	/* check if the MPMR is filled */
+	if (caam_hal_ctrl_is_mp_set(ctrl_addr))
+		is_filled = true;
+
+	DMSG("is_filled = %s", is_filled ? "true" : "false");
+
+	if (!is_filled) {
+		/*
+		 * Fill the MPMR with the most significant input value and
+		 * complete with 0's if value too short.
+		 */
+		min_size = MIN(msg_mpmr->length, (size_t)MPMR_NB_REG);
```

Comment:
cast needed?

Reply:
yes `warning: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]`